### PR TITLE
390 test overlay physical barriers

### LINKED ISSUE #390: Add tests for `cluster.overlay_gdf_physical_barriers`

### PR DESCRIPTION
**Pull request by crispy-wonton**
Fixes #390 

# Description

Test the `cluster.overlay_gdf_physical_barriers` function. There is one test that tests for different cases of barrier intersection with building footprints.

New files
- `asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet` - small parquet file containing the test Voronoi polygons that I created for use in testing. This is fictional data. I saved as a separate file as it was quite a large geodataframe to have written in the code.

Modified files:
- `asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py` - add new fixtures and test required to test the overlay function.

# Instructions for Reviewer

In order to test the code in this PR you need to ...
run the following code in terminal:
```
pip install pytest
pytest
```
There are 8 tests (1 new to this PR) which should all pass.

Please pay special attention to ...
- Are there any test cases missing of ways that the barrier polygons could intersect with the building footprints?
- Are there any other tests we should perform on this function specifically that aren't covered by the `generate_gdf_cluster` tests?

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 1]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+            "voronoi_polygons_test_set.parquet",
+        )
+        return gpd.read_parquet(fpath).set_crs(epsg=27700)
+
+    @pytest.fixture(scope="class")
+    def gdf_line_overlay_mixed_buildings(self):
+        """
+        Create a geodataframe of line polygons (linestrings converted to polygons) that meet the following criteria:
+        1. One road separates the corner of a building off from the rest of the footprint.
+        2. One road bisects a building.
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:628]: Why are we adding this one? Does this happen in reality?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 2]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+        """
+        fpath = os.path.join(
+            PROJECT_DIR,
+            "asf_heat_pump_suitability/pipeline/cluster/tests",
+            "voronoi_polygons_test_set.parquet",
+        )
+        return gpd.read_parquet(fpath).set_crs(epsg=27700)
+
+    @pytest.fixture(scope="class")
+    def gdf_line_overlay_mixed_buildings(self):
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:624]: so this is to mimic roads?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 3]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+                    ]
+                ),
+            ]
+        }
+
+        # Dissolve overlapping polygons and explode back to one line per polygon
+        return gpd.GeoDataFrame(line_dict, crs="EPSG:27700").dissolve().explode()
+
+    @pytest.fixture(scope="class")
+    def gdf_polygon_overlay_mixed_buildings(self):
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:685]: and this is to mimic greenspaces, commercial buildindgs, etc.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 4]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+    @pytest.fixture(scope="class")
+    def gdf_line_overlay_mixed_buildings(self):
+        """
+        Create a geodataframe of line polygons (linestrings converted to polygons) that meet the following criteria:
+        1. One road separates the corner of a building off from the rest of the footprint.
+        2. One road bisects a building.
+        3. The other roads go around and between buildings to separate clusters.
+        The buildings affected are separate from those affected by the overlay polygons.
+        """
+        line_dict = {
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:632]: i'm guessing you tried to mimic the typical sizes of roads

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 5]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+        # Dissolve overlapping polygons and explode back to one line per polygon
+        return gpd.GeoDataFrame(line_dict, crs="EPSG:27700").dissolve().explode()
+
+    @pytest.fixture(scope="class")
+    def gdf_polygon_overlay_mixed_buildings(self):
+        """
+        Create a geodataframe of overlay polygons that meet the following criteria:
+        1. Overlaps with the corner of a building.
+        2. Completely overlaps a building.
+        3. Bisects a building.
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:690]: I can't find a use case for this in my mind?

Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:690]: and we're saying building but it's not really building right? it's cluster

voronois -> clusters -> we need to make wholes in these where parks, commercial buildings, etc exist. right?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 6]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+            line_overlay_gdf=gdf_line_overlay_mixed_buildings,
+            polygon_overlay_gdf=gdf_polygon_overlay_mixed_buildings,
+            id_col="building_id",
+        )
+
+        results = cells_gdf["building_id"]
+        expected = domestic_tech_gdf["building_id"]
+        assert set(results) == set(
+            expected
+        ), "Some buildings do not have a cell after overlaying barriers"
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:789]: Again, maybe I'm just misremembering but does this happen after we have a cell for a building or after merging the cells?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 7]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+        ), f"Number of Voronoi cells {len(cells_gdf)} != number of domestic buildings {len(domestic_tech_gdf)} after overlay"
+
+        # Check if the uncontained area of building footprints is effectively zero (e.g., less than 1 square millimeter)
+        # This is to account for floating point errors which can cause tiny slivers of building not to be covered by
+        # the cells.
+        uncontained_slivers_gdf = domestic_tech_gdf.overlay(
+            cells_gdf, how="difference", keep_geom_type=False
+        )
+        uncontained_slivers_gdf["area"] = uncontained_slivers_gdf["geometry"].area
+        results = uncontained_slivers_gdf[uncontained_slivers_gdf["area"] > 1e-5]
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:802]: is it worth checking how much this is happening, as a proportion? maybe npt?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py [hunk 8]
```diff
@@ -572,5 +576,206 @@ def test_clip_voronoi_to_buffer(
...
+
+        assert len(cells_gdf) == len(
+            domestic_tech_gdf
+        ), f"Number of Voronoi cells {len(cells_gdf)} != number of domestic buildings {len(domestic_tech_gdf)} after overlay"
+
+        # Check if the uncontained area of building footprints is effectively zero (e.g., less than 1 square millimeter)
+        # This is to account for floating point errors which can cause tiny slivers of building not to be covered by
+        # the cells.
+        uncontained_slivers_gdf = domestic_tech_gdf.overlay(
+            cells_gdf, how="difference", keep_geom_type=False
```
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:799]: why keep_geom_type=False?

File: asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet
Comment (sofiapinto) [inline asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet:1]: To note that i haven't inspected this file

**Review by sofiapinto — COMMENTED** [review]
Thank you for this wonderful work @crispy-wonton - have mainly left questions below. Haven't been able to run the code (because of having deleted the conda env and now using uv, but uv isn't setup here). Can we bring the latest dev changes into 384_test_clustering and then the latest changes from 384_test_clustering into this branch390_test_overlay_physical_barriers so that i am able to run the tests?

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 1 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:628]: It is rare but it has happened, in I think one edge case that I have seen in Plymouth so far.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 2 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:624]: Yes or other things like rivers.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 4 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:632]: Yes, it is meant to be 3m wide. It doesn't matter too much here though as even a very unrealistically small width (e.g. 1cm) will still separate the clusters which is what the tests are for here.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 5 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:690]: Not sure why this happens exactly but it does. I think it might be rare cases where a road goes under a building (e.g. pic) or it could be errors.

Image

Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:690]: This is for testing a function that deals with the Voronoi cells of individual buildings. In the cases I've listed in the docstring, I specifically mean building, because this affects the final cluster shape. The function needs to be able to handle cases where the overlaying roads / green spaces etc. overlap with a building footprint. Because the final cluster polygon must contain the whole building footprint regardless.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 6 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:789]: This is testing the `overlay_gdf_physical_barriers` function which operates at the level of individual cells containing a single domestic building. This is before merging cells.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 8 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:799]: Fair question, it retains any point or linestring geometries created by the overlay, as well as polygons. I think it's irrelevant as the area will be zero so really it doesn't matter if set to True / False. I think I probably set it to False to suppress a warning about dropped geometries.

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 7 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:802]: Do you mean in the true dataset? I don't think it's worth checking in the test set.

File: asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet:1]: It might be worth just loading and displaying it if you haven't already so that you can visualise the test set of building footprints.

```
import geopandas as gpd
gdf = gpd.read_parquet("asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet")

gdf.plot()
```

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 5 above)
Comment (sofiapinto) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:690]: ahhh makes so much sense, thank you!

File: asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py (hunk 7 above)
Comment (sofiapinto) [reply asf_heat_pump_suitability/pipeline/cluster/tests/test_cluster.py:802]: yep in the true dataset - but maybe not?

File: asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet
Comment (sofiapinto) [reply asf_heat_pump_suitability/pipeline/cluster/tests/voronoi_polygons_test_set.parquet:1]: Have now inspected the file :)

**Review by sofiapinto — DISMISSED** [review]

**Review by sofiapinto — APPROVED** [review]
